src/hotspot/share/oops/array.hpp

```diff
@@ -91,30 +91,30 @@ class Array: public MetaspaceObj {
   Array(int length, T init) : _length(length) {
     assert(length >= 0, "illegal length");
     for (int i = 0; i < length; i++) {
-      _data[i] = init;
+      data()[i] = init;
     }
   }
 
  public:
 
   // standard operations
   int  length() const                 { return _length; }
-  T* data()                           { return _data; }
+  T* data() const                     { return reinterpret_cast<T*>(reinterpret_cast<uintptr_t>(this) + base_offset_in_bytes()); }
```

Comment:
Adding the const-qualifier to the `data()` function and then implicitly
casting it away (by casting through intptr_t) is wrong.  Either don't
const-qualify (and leave it to some future use that needs such to address
appropriately), or have two functions.  Also, the line length is excessive.
So this:

```
T* data() {
  return reinterpret_cast<T*>(
    reinterpret_cast<char*>(this) + base_offset_in_bytes());
}
```
and optionally add this:
```
const T* data() const {
  return reinterpret_cast<const T*>(
    reinterpret_cast<const char*>(this) + base_offset_in_bytes());
}
```

Reply:
Thanks a lot @kimbarrett !

I updated around stringop-overflow warning in jfrTraceIdBits.inline.hpp , and added two `data()` in `Array` class. They works fine on my GCC 12 on Fedora 36. Could you review again?